independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/handlers/FormBodyHandler.java

```diff
@@ -3,60 +3,129 @@
 
-    public FormBodyHandler(boolean alsoSetInputStream) {
+    public FormBodyHandler(boolean alsoSetInputStream, Supplier<Executor> executorSupplier) {
         this.alsoSetInputStream = alsoSetInputStream;
+        this.executorSupplier = executorSupplier;
+    }
+
+    @Override
+    public void configure(RuntimeConfiguration configuration) {
+        formParserFactory = FormParserFactory.builder(false, executorSupplier)
+                .addParser(new MultiPartParserDefinition(executorSupplier)
+                        .setFileSizeThreshold(0)
+                        .setMaxAttributeSize(configuration.limits().maxFormAttributeSize())
+                        .setMaxEntitySize(configuration.limits().maxBodySize().orElse(-1L))
+                        .setDeleteUploadsOnEnd(configuration.body().deleteUploadedFilesOnEnd())
+                        .setTempFileLocation(Path.of(configuration.body().uploadsDirectory())))
+                .addParser(new FormEncodedDataDefinition()
+                        .setMaxAttributeSize(configuration.limits().maxFormAttributeSize()))
+                .build();
+
+        try {
+            Files.createDirectories(Paths.get(configuration.body().uploadsDirectory()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
     }
 
     @Override
     public void handle(ResteasyReactiveRequestContext requestContext) throws Exception {
         // in some cases, with sub-resource locators or via request filters, 
         // it's possible we've already read the entity
-        if (requestContext.hasInputStream()) {
+        if (requestContext.getFormData() != null) {
             // let's not set it twice
             return;
         }
         ServerHttpRequest serverHttpRequest = requestContext.serverRequest();
-        if (serverHttpRequest.isRequestEnded()) {
+        if (BlockingOperationSupport.isBlockingAllowed()) {
+            //blocking IO approach
+
+            FormDataParser factory = formParserFactory.createParser(requestContext);
+            if (factory == null) {
+                return;
+            }
+            CapturingInputStream cis = null;
             if (alsoSetInputStream) {
-                // do not use the EmptyInputStream.INSTANCE marker
-                requestContext.setInputStream(new ByteArrayInputStream(NO_BYTES));
+                // the TCK allows the body to be read as a form param and also as a body param
+                // the spec is silent about this
+                // TODO: this is really really horrible and hacky and needs to be fixed.
+                cis = new CapturingInputStream(requestContext.getInputStream());
+                requestContext.setInputStream(cis);
+            }
+            factory.parseBlocking();
+            if (alsoSetInputStream) {
+                requestContext.setInputStream(new ByteArrayInputStream(cis.baos.toByteArray()));
+            }
+        } else if (alsoSetInputStream) {
```

Comment:
This case will switch to a worker thread, right? So that's going to be expensive, but I'm not sure I remember what triggers this case of `alsoSetInputStream`. If it's just spec TCK edge-cases then I don't care.

Reply:
It's TCK spec edge cases AFAIK, when you want to both parse the body as a form and inject the raw body as a string. 

I don't think the user visible parts changed? Which parts are you talking about?

Reply:
OK thanks.